Content.Server/Fluids/EntitySystems/MoppingSystem.cs

```diff
@@ -0,0 +1,363 @@
+using Robust.Shared.Player;
+using Robust.Shared.Map;
+using JetBrains.Annotations;
+
+namespace Content.Server.Fluids.EntitySystems;
+
+[UsedImplicitly]
+public sealed class MoppingSystem : EntitySystem
+{
+        [Dependency] private readonly DoAfterSystem _doAfterSystem = default!;
+        [Dependency] private readonly SpillableSystem _spillableSystem = default!;
+        [Dependency] private readonly TagSystem _tagSystem = default!;
+        [Dependency] private readonly IMapManager _mapManager = default!;
+        [Dependency] private readonly SolutionContainerSystem _solutionSystem = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<AbsorbentComponent, AfterInteractEvent>(OnAfterInteract);
+        SubscribeLocalEvent<TransferCancelledEvent>(OnTransferCancelled);
+        SubscribeLocalEvent<TransferCompleteEvent>(OnTransferComplete);
+    }
+
+    private void OnAfterInteract(EntityUid uid, AbsorbentComponent component, AfterInteractEvent args)
+    {
+        var user = args.User;
+        var used = args.Used;
+        var target = args.Target;
+
+        _solutionSystem.TryGetSolution(used, "absorbed", out var absorbedSolution);
+
+        var toolAvailableVolume = FixedPoint2.New(0);
+        var toolCurrentVolume = FixedPoint2.New(0);
+        var transferAmount = FixedPoint2.New(0);
+
+        if (absorbedSolution is not null)
+        {
+            toolAvailableVolume = absorbedSolution.AvailableVolume;
+            toolCurrentVolume = absorbedSolution.CurrentVolume;
+        }
+
+        if (!args.CanReach)
+        {
+            return;
+        }
+
+        // For adding liquid to an empty floor tile
+        if (target is null // if a tile is clicked
+            && !args.Handled)
+        {
+            ReleaseToFloor(args.ClickLocation, component, absorbedSolution);
+            args.Handled = true;
+            return;
+        }
+        else if (target is not null)
+        {
+            // Handle our do_after logic
+            HandleDoAfter(user, used, target.Value, component, toolCurrentVolume, toolAvailableVolume);
+        }
+
+        args.Handled = true;
+        return;
+    }
+
+
+    private void ReleaseToFloor(EntityCoordinates clickLocation, AbsorbentComponent absorbent, Solution? absorbedSolution)
+    {
+
+        if ((_mapManager.TryGetGrid(clickLocation.GetGridId(EntityManager), out var mapGrid)) // needs valid grid
+            && absorbedSolution != null) // needs a solution to place on the tile
+        {
+            TileRef tile = mapGrid.GetTileRef(clickLocation);
+
+            // Drop some of the absorbed liquid onto the ground
+            var releaseAmount = FixedPoint2.Min(absorbent.ResidueAmount, absorbedSolution.CurrentVolume);
+            var releasedSolution = _solutionSystem.SplitSolution(absorbent.Owner, absorbedSolution, releaseAmount);
+            _spillableSystem.SpillAt(tile, releasedSolution, "PuddleSmear");
```

Comment:
maybe have the puddle prototype ID defined on the absorbent component? idk maybe thats a lil useless

Reply:
Ah, I see what you mean, and agree that it's ambiguous where it truly belongs. Heck, maybe some default would belong under SpillAt?

My #6848 now handles random state selection and rotation, so I'd be fine with leaving PuddleSmear hardcoded under MoppingSolution as I have it here.

Would it be cleaner if I made it a `const` near the beginning of `MoppingSystem` maybe?

Reply:
Made it a `const` within `MoppingSystem`.